xcpmd/src/battery.c

```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
         return NORMAL;
 }
 
+
+//Returns whether a battery is charging, discharging, full, etc.
+int get_battery_charge_state(unsigned int battery_index) {
+
+    unsigned int percent, i;
+
+    if (last_status[battery_index].state & 0x1)
+        return BATT_DISCHARGING;
+    else if (last_status[battery_index].state & 0x2)
+        return BATT_CHARGING;
+    else {
+        /* We're not charging nor discharging... */
+        percent = get_battery_percentage(battery_index);
+        /* Are we full or empty? */
+        if (percent > 90)
+            return BATT_FULL;
+        else if (percent < 10)
+             return BATT_EMPTY;
+        else {
+            /* Is anybody else (dis)charging? */
+            for (i = 0; i < num_battery_structs_allocd; ++i) {
+                if (i != battery_index &&
+                    last_status[i].present == YES &&
+                    (last_status[i].state & 0x1 || last_status[i].state & 0x2)) {
+                    break;
+                }
+            }
+            if (i < num_battery_structs_allocd) {
+                /* Yes! */
+                /* If the other battery is charging, we're pending charge */
+                if (last_status[i].state & 0x2)
+                    return BATT_PENDING_CHARGE;
+                /* If the other battery is discharging, we're pending discharge */
+                else if (last_status[i].state & 0x1)
+                    return BATT_PENDING_DISCHARGE;
+            } else
```

Comment:
This was moved from xcpmd_dbus_server.c, and I was trying not to modify code that wasn't mine any more than strictly necessary, but I can remove that if you prefer.